README.md
[![Build Status](https://travis-ci.org/nramon/madrid.air.svg?branch=master)](https://travis-ci.org/nramon/madrid.air) [![codecov](https://codecov.io/gh/nramon/madrid.air/branch/master/graph/badge.svg)](https://codecov.io/gh/nramon/madrid.air) [![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/265/badge)](https://bestpractices.coreinfrastructure.org/projects/265)

# Madrid Air Quality Data Parser

Convert raw air quality data published by [http://datos.madrid.es/](http://datos.madrid.es/) to a clean data frame.

The original format is hard to interpret:

```
2807900214060201042844.55V52.04V48.30V48.37V23.60V27.89V28.42V13.75V19.11V38.89V26.32V27.05V41.60V46.91V58.52V77.15V74.55V60.91V52.33V46.20V23.47V9.500V12.24V18.92V
```

[madrid.air](https://github.com/nramon/madrid.air) allows you to convert it to a more readable form:

```
     Station Year Month Day Hour Magnitude Value
1	28079002 2001     4  28    1        O3 44.55
2	28079002 2001     4  28    2        O3 52.04
3	28079002 2001     4  28    3        O3 48.30
...
```

## Installation

To install [madrid.air](https://github.com/nramon/madrid.air) from GitHub start the R CMD interface and run the following commands:

```
install.packages("devtools")
library(devtools)
install_github("nramon/madrid.air")
```

## Downloading Madrid air quality data

Both [daily](http://datos.madrid.es/sites/v/index.jsp?vgnextoid=aecb88a7e2b73410VgnVCM2000000c205a0aRCRD&vgnextchannel=374512b9ace9f310VgnVCM100000171f5a0aRCRD) and [hourly](http://datos.madrid.es/sites/v/index.jsp?vgnextoid=f3c0f7d512273410VgnVCM2000000c205a0aRCRD&vgnextchannel=374512b9ace9f310VgnVCM100000171f5a0aRCRD) data can be freely downloaded from [http://datos.madrid.es/](http://datos.madrid.es/).

## Usage

See the package documentation for usage instructions and examples:

```
library(madrid.air)
?madrid.air
?madrid.air.parse
?madrid.air.magnitudes
?madrid.air.stations
```

## Bugs and issues

To provide feedback please use the [issue tracker](https://github.com/nramon/madrid.air/issues).

## Contributing

See [CONTRIBUTING.md](https://github.com/nramon/madrid.air/blob/master/CONTRIBUTING.md).

## Copyright

Copyright (C) 2016 Ramon Novoa <ramonnovoa@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log

## v1.0.0 (2016-12-15)

 Initial release.


tests/testthat.R
library(testthat)
library(madrid.air)

test_check("madrid.air")


tests/testthat/test.lintr.R
# Copyright (C) 2016 Ramon Novoa <ramonnovoa AT gmail DOT com>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# See: https://github.com/jimhester/lintr
if (requireNamespace("lintr", quietly = TRUE)) {
  context("lints")
  test_that("Package Style", {
    lintr::expect_lint_free()
  })
}


tests/testthat/test.parse.hourly.R
# Copyright (C) 2016 Ramon Novoa <ramonnovoa AT gmail DOT com>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

test_that("hourly data is properly parsed", {

  # Test daily data for 2001/04/28
  data <- madrid.air.parse("test.parse.hourly.data")

  # One row per hour
  expect_equal(nrow(data), 24)

  # Seven variables.
  expect_equal(ncol(data), 7)

  # Test variables for 2001/04/28 12:00
  expect_equal(as.character(data[12, 1]), "28079002")
  expect_equal(as.character(data[12, 2]), "2001")
  expect_equal(as.character(data[12, 3]), "4")
  expect_equal(as.character(data[12, 4]), "28")
  expect_equal(as.character(data[12, 5]), "12")
  expect_equal(as.character(data[12, 6]), "O3")
  expect_equal(data[12, 7], 27.05)
})


R/madrid.air.R
# Copyright (C) 2016 Ramon Novoa <ramonnovoa AT gmail DOT com>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

#' @author Ramon Novoa \email{ramonnovoa@@gmail.com}
#' @section Functions:
#' \code{\link{madrid.air.parse}}
#' @section Datasets:
#' \code{\link{madrid.air.magnitudes}}
#' \code{\link{madrid.air.stations}}
#' @seealso \url{https://github.com/nramon/madrid.air}
"_PACKAGE"


data-raw/madrid.air.stations.R
# Copyright (C) 2016 Ramon Novoa <ramonnovoa AT gmail DOT com>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

################################################################################
# Stations from: http://datos.madrid.es/FWProjects/egob/contenidos/datasets/fich
# eros/MedioAmbiente_CalidadAire/INTPHORA-DIA_V2.2.pdf
# Coordinates from: http://www.eea.europa.eu/data-and-maps/data/airbase-the-euro
# pean-air-quality-database-8
################################################################################
madrid.air.stations <- data.frame(
  rbind(
    c("28079001", "Pº. Recoletos", -3.691945, 40.422501,
      "Closed on 04/05/2009"),
    c("28079002", "Glta. de Carlos V", -3.691945, 40.409725,
      "Closed on 04/12/2006"),
    c("28079004", "Pza. de España", -3.712222, 40.424167, ""),
    c("28079006", "Pza. Dr. Marañón", -3.690833, 40.437778,
      "Closed on 27/11/2009"),
    c("28079007", "Pza. M. de Salamanca", -3.680278, 40.421669,
      "Closed on 30/12/2009"),
    c("28079008", "Escuelas Aguirre", -3.682222, 40.421667, ""),
    c("28079009", "Pza. Luca de Tena", -3.693611, 40.402225,
      "Closed on 07/12/2009"),
    c("28079011", "Av. Ramón y Cajal", -3.677222, 40.451667, ""),
    c("28079012", "Pza. Manuel Becerra", -3.668611, 40.428612,
      "Closed on 30/12/2009"),
    c("28079014", "Pza. Fdez. Ladreda", -3.716667, 40.385,
      "Closed on 02/12/2009"),
    c("28079015", "Pza. Castilla", -3.688611, 40.468334,
      "Closed on 17/10/2008"),
    c("28079016", "Arturo Soria", -3.639167, 40.44, ""),
    c("28079017", "Villaverde Alto", -3.705, 40.346944, ""),
    c("28079018", "C/ Farolillo", -3.731944, 40.394722, ""),
    c("28079023", "Final C/ Alcalá", -3.604444, 40.450279,
      "Closed on 30/12/2009"),
    c("28079024", "Casa de Campo", -3.749167, 40.42, ""),
    c("28079025", "Santa Eugenia", -3.6025, 40.379444,
      "Closed on 16/11/2009"),
    c("28079026", "Urb. Embajada (Barajas)", -3.578889, 40.459999,
      "Closed on 11/01/2010"),
    c("28079027", "Barajas", -3.575, 40.474722, ""),
    c("28079035", "Pza. del Carmen", -3.703333, 40.419167, ""),
    c("28079036", "Moratalaz", -3.645278, 40.408056, ""),
    c("28079038", "Cuatro Caminos", -3.706389, 40.445278, ""),
    c("28079039", "Barrio del Pilar", -3.711389, 40.478333, ""),
    c("28079040", "Vallecas", -3.651389, 40.388056, ""),
    c("28079047", "Méndez Álvaro", -3.686667, 40.398056,
      "Since 21/12/2009"),
    c("28079048", "Pº. Castellana", -3.690278, 40.439722,
      "Since 01/06/2010"),
    c("28079049", "Retiro", -3.685556, 40.409722,
      "Since 01/01/2010"),
    c("28079050", "Pza. Castilla", -3.688333, 40.465833,
      "Since 08/02/2010"),
    c("28079054", "Ensanche Vallecas", -3.611944, 40.372778,
      "Since 11/12/2009"),
    c("28079055", "Urb. Embajada (Barajas)", -3.580556, 40.4625,
      "Since 20/01/2010"),
    c("28079056", "Pza. Fdez. Ladreda", -3.718611, 40.384722,
      "Since 18/01/2010"),
    c("28079057", "Sanchinarro", -3.660278, 40.494167,
      "Since 24/11/2009"),
    c("28079058", "El Pardo", -3.774444, 40.518056,
      "Since 30/11/2009"),
    c("28079059", "Parque Juan Carlos I", -3.608889, 40.465,
      "Since 14/12/2009"),
    c("28079060", "Tres Olivos", -3.689722, 40.500556,
      "Since 14/01/2010")
  ),
  stringsAsFactors = FALSE
)

# Set proper column names and fix column types.
names(madrid.air.stations) <- c("Station", "Location", "Longitude", "Latitude",
                                "Comments")
madrid.air.stations$Longitude <- as.double(madrid.air.stations$Longitude)
madrid.air.stations$Latitude <- as.double(madrid.air.stations$Latitude)

# Set the right encoding for the locations.
Encoding(madrid.air.stations$Location) <- "UTF-8"


data-raw/madrid.air.magnitudes.R
# Copyright (C) 2016 Ramon Novoa <ramonnovoa AT gmail DOT com>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

################################################################################
# Magnitudes from: http://datos.madrid.es/FWProjects/egob/contenidos/datasets/fi
# cheros/MedioAmbiente_CalidadAire/INTPHORA-DIA_V2.2.pdf
################################################################################
madrid.air.magnitudes <- data.frame(
    rbind(
        c("Benzene", "BEN", 30, "µg/m³"),
        c("Hydrocarbons (methane)", "CH4", 43, "mg/m³"),
        c("Carbon monoxide", "CO", 6, "mg/m³"),
        c("Wind direction", "DV", 82, "º or quadrant"),
        c("Ethylbenzene", "EBE", 35, "µg/m³"),
        c("Relative humidity", "HR", 86, "%"),
        c("Acid Rain", "LLA", 89, "pH"),
        c("Precipitation", "LL", 92, "l/m²"),
        c("Metaxylene", "MXY", 37, "µg/m³"),
        c("Non-methane hydrocarbons (hexane)", "NMHC", 44, "mg/m³"),
        c("Nitrogen dioxide", "NO2", 7, "µg/m³"),
        c("Nitrogen monoxide", "NO", 8, "µg/m³"),
        c("Nitrogen oxides", "NOx", 12, "µg/m³"),
        c("Ozone", "O3", 14, "µg/m³"),
        c("Orthoxylene", "OXY", 39, "µg/m³"),
        c("Particulate matter < 10 µm", "PM10", 10, "µg/m³"),
        c("Particulate matter < 2.5 µm", "PM2.5", 9, "µg/m³"),
        c("Pressure", "PRB", 87, "mb"),
        c("Paraxylene", "PXY", 38, "µg/m³"),
        c("Solar irradiance", "RS", 88, "kW/m²"),
        c("Sulfur dioxide", "SO2", 1, "µg/m³"),
        c("Total hydrocarbons (hexane)", "TCH", 42, "mg/m³"),
        c("Temperature", "TMP", 83, "ºC"),
        c("Toluene", "TOL", 20, "µg/m³"),
        c("Ultraviolet radiation", "UV", 80, "mW/m²"),
        c("Wind speed", "VV", 81, "m/s"),
        c("Unknown", "Unknown", 85, "Unknown")
    ),
	stringsAsFactors = FALSE
)

# Set proper column names.
names(madrid.air.magnitudes) <- c("Name", "Abbreviation", "Magnitude", "Unit")

# Set the right encoding for magnitude names and units.
Encoding(madrid.air.magnitudes$Name) <- "UTF-8"
Encoding(madrid.air.magnitudes$Unit) <- "UTF-8"


R/madrid.air.parse.R
# Copyright (C) 2016 Ramon Novoa <ramonnovoa AT gmail DOT com>
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# We will need the magnitude and station datasets.
data(madrid.air.magnitudes, envir = environment())
data(madrid.air.stations, envir = environment())

# Supress warnings.
Month <- NULL
Day <- NULL
Hour <- NULL
Value <- NULL
ValidationCode <- NULL

################################################################################
#' Air quality data parser.
#'
#' Parses raw air quality data published by \url{http://datos.madrid.es/} and returns
#' a clean data frame.
#'
#' @param input Input file containing the raw air quality data.
#' @param output Optional CSV file where the air quality data data frame will
#' be saved.
#' @param station Filter data by station code (see
#' \code{\link{madrid.air.stations}}).
#' @param magnitude Filter data by magnitude name (see
#' \code{\link{madrid.air.magnitudes}}).
#' @return A data frame containing air quality data.
#' @examples
#' # Download and parse daily data.
#' download.file('http://datos.madrid.es/egob/catalogo/201410-14-calidad-aire-diario.txt',
#'               '201410-14-calidad-aire-diario.txt')
#' madrid.air.parse('201410-14-calidad-aire-diario.txt')
#'
#' # Save air quality data to disk.
#' madrid.air.parse('201410-14-calidad-aire-diario.txt',
#'                  output='201410-14-calidad-aire-diario.csv')
#'
#' # Download and parse hourly data.
#' download.file('http://datos.madrid.es/egob/catalogo/201200-29-calidad-aire-horario.zip',
#'               '201200-29-calidad-aire-horario.zip')
#' unzip('201200-29-calidad-aire-horario.zip')
#' madrid.air.parse('Ene_mo01.txt')
#'
#' # Check station information.
#' data(madrid.air.stations)
#' madrid.air.stations
#'
#' # Filter by station.
#' madrid.air.parse('Ene_mo01.txt', station=28079001)
#'
#' # Check magnitude information.
#' data(madrid.air.magnitudes)
#' madrid.air.magnitudes
#'
#' # Filter by magnitude.
#' madrid.air.parse('Ene_mo01.txt', magnitude='CO')
#' @export
################################################################################
madrid.air.parse <- function (input,
                              output = NA,
                              station = NA,
                              magnitude = NA) {

  # Read the first record to parse the year and period.
  air_data <- utils::read.fwf(input, widths = c(8, 2, 2, 2, 2, 2, 2),
                              stringsAsFactors = FALSE, n = 1)
  period <- air_data[1, 4]
  year <- air_data[1, 5] + 2000

  # Parse air quality data.
  if (period == 2) {
    air_data <- madrid.air.parser.hourly(input)
  } else if (period == 4) {
    air_data <- madrid.air.parser.daily(input, year)
  } else {
    stop("Invalid file format.")
  }

  # Filter by station.
  if (!is.na(station)) {
    air_data <- air_data[air_data$Station == station, ]
  }

  # Filter by magnitude
  if (!is.na(magnitude)) {
    air_data <- air_data[air_data$Magnitude == magnitude, ]
  }

  # Save the dataset to disk.
  if (!is.na(output)) {
    utils::write.csv(air_data, file = output, row.names = F, quote = F)
  }

  air_data
}

################################################################################
################################################################################
## Helper functions.
################################################################################
################################################################################

################################################################################
# Parses daily air quality data and returns a data frame.
#
# Args:
#   input: Source file containing the raw air quality data.
#   year: The year the data was collected.
#
# Returns:
#   A data frame.
################################################################################
madrid.air.parser.daily <- function (input, year, output = NA) {

  # There was an extra column before 2011 with the value 00 in daily
  # data.
  if (year < 2011) {
    air_data <- utils::read.fwf(input, widths = c(8, 2, 2, 2, 2, 2, 2,
                                rep(6, 31)), stringsAsFactors = FALSE)

    # Remove the extra column.
    air_data <- air_data[, -7]
  } else {
    air_data <- utils::read.fwf(input, widths = c(8, 2, 2, 2, 2, 2,
                                rep(6, 31)), stringsAsFactors = FALSE)
  }

  # Remove the measurement technique and period columns.
  air_data <- air_data[, -c(3, 4)]

  # Set appropriate column names.
  names(air_data) <- c("Station", "Magnitude", "Year", "Month", c(1:31))

  # Leave one daily observation per row.
  air_data <- tidyr::gather(air_data, Day, Value, 5:35)

  # Perform common parsing operations.
  air_data <- madrid.air.parser.common(air_data)

  # Rearrange the columns.
  air_data <- air_data[, c("Station", "Year", "Month", "Day", "Magnitude",
                          "Value")]

  # Sort by date.
  air_data <- dplyr::arrange(air_data, Month, Day)

  air_data
}

################################################################################
# Parses hourly air quality data and returns a data frame.
#
# Args:
#   input: Source file containing the raw air quality data.
#
# Returns:
#   A data frame.
################################################################################
madrid.air.parser.hourly <- function (input) {

  air_data <- utils::read.fwf(input,
                              widths = c(8, 2, 2, 2, 2, 2, 2, rep(6, 24)),
                              stringsAsFactors = FALSE)

  # Remove the measurement technique and period columns.
  air_data <- air_data[, -c(3, 4)]

  # Set appropriate column names.
  names(air_data) <- c("Station", "Magnitude", "Year", "Month", "Day",
                       c(1:24))

  # Leave one hourly observation per row.
  air_data <- tidyr::gather(air_data, Hour, Value, 6:29)

  # Perform common parsing operations.
  air_data <- madrid.air.parser.common(air_data)

  # Rearrange the columns.
  air_data <- air_data[, c("Station", "Year", "Month", "Day", "Hour",
                          "Magnitude", "Value")]

  # Fix some column types.
  air_data$Hour <- as.integer(air_data$Hour)

  # Sort by date.
  air_data <- dplyr::arrange(air_data, Month, Day, Hour)

  air_data
}

################################################################################
# Performs parsing operation common to daily and hourly data.
#
# Args:
#   input: A data frame containing daily or hourly air quality data.
#
# Returns:
#   A data frame.
################################################################################
madrid.air.parser.common <- function (air_data) {

  # Separate the validation code from the actual value.
  air_data <- tidyr::separate(air_data, Value, c("Value", "ValidationCode"),
                              sep = 5)

  # We will only keep valid data (validation code "V").
  air_data <- dplyr::filter(air_data, ValidationCode == "V")

  # Sanity check.
  if (nrow(air_data) == 0) {
    stop("Invalid file format.")
  }

  # Remove the validation code from the dataset.
  air_data <- air_data[, -which(names(air_data) %in% c("ValidationCode"))]

  # Convert values to double precision numbers.
  air_data$Value <- as.double(air_data$Value)

  # Save magnitudes as a factor.
  air_data$Magnitude <- factor(air_data$Magnitude,
                               levels = madrid.air.magnitudes[, "Magnitude"],
                               labels = madrid.air.magnitudes[, "Abbreviation"])

  # Remove unknown magnitudes.
  # Note: There is no description for magnitude 85 in the docs!!!
  air_data <- air_data[!is.na(air_data$Magnitude), ]

  # Some station codes changed after 2011. Rename old codes to new codes.
  # Pza. del Carmen.
  air_data[air_data$Station == 28079003, "Station"] <- 28079035

  # Barrio del Pilar.
  air_data[air_data$Station == 28079005, "Station"] <- 28079039

  # Cuatro Caminos.
  air_data[air_data$Station == 28079010, "Station"] <- 28079038

  # Vallecas.
  air_data[air_data$Station == 28079013, "Station"] <- 28079040

  # Moratalaz.
  air_data[air_data$Station == 28079020, "Station"] <- 28079036

  # Tres Olivos.
  air_data[air_data$Station == 28079086, "Station"] <- 28079060

  # Fix the Year column.
  air_data$Year <- air_data$Year + 2000

  # Fix some column types.
  air_data$Year <- as.integer(air_data$Year)
  air_data$Month <- as.integer(air_data$Month)
  air_data$Day <- as.integer(air_data$Day)
  air_data$Station <- as.factor(air_data$Station)

  air_data
}
